### PR TITLE
Fix join key nullable check for anti join types

### PR DESCRIPTION
**Pull request by xiaoxmeng**
The current anti join handling is missing join key
nullable check if the build side is empty and it has
join filtering.
Add unit tests to reproduce and verified the fix.

### CODE REVIEW

File: velox/exec/HashProbe.cpp [hunk 1]
```diff
@@ -267,7 +267,11 @@ void HashProbe::addInput(RowVectorPtr input) {
     VELOX_CHECK(
         isAntiJoins(joinType_) || isLeftJoin(joinType_) ||
         isFullJoin(joinType_));
-    return;
+    if (!isAntiJoins(joinType_)) {
+      return;
+    }
+    // For ant join types we need to decode the join keys columns to find out
```
Comment (mbasmanova) [inline velox/exec/HashProbe.cpp:273]: typo: ant -> anti

File: velox/exec/HashProbe.cpp [hunk 2]
```diff
@@ -280,6 +284,11 @@ void HashProbe::addInput(RowVectorPtr input) {
 
   deselectRowsWithNulls(hashers_, nonNullInputRows_);
 
+  if (table_->numDistinct() == 0) {
```
Comment (mbasmanova) [inline velox/exec/HashProbe.cpp:287]: I'm not sure I understand this fix. Would you explain?

**Review by Yuhta — APPROVED** [review]

File: velox/exec/HashProbe.cpp (hunk 2 above)
Comment (mbasmanova) [reply velox/exec/HashProbe.cpp:287]: > when accessed later

Can you point me to that code? Does this happen only if there is an extra filter? If so, can be skip decoding inputs if there is no filter?

Comment (Yuhta) [reply velox/exec/HashProbe.cpp:287]: Yes it's only happen with filter

Comment (mbasmanova) [reply velox/exec/HashProbe.cpp:287]: Got it. Would it make sense to do the extra processing (decoding of the input) only when the filter is present?

File: velox/exec/HashProbe.cpp [hunk 3]
```diff
@@ -267,7 +267,11 @@ void HashProbe::addInput(RowVectorPtr input) {
     VELOX_CHECK(
         isAntiJoins(joinType_) || isLeftJoin(joinType_) ||
         isFullJoin(joinType_));
-    return;
+    if (!isAntiJoins(joinType_)) {
```
[the remark this answers is not in the document]
Comment (xiaoxmeng) [reply velox/exec/HashProbe.cpp:270]: I think this handling is for isAntiJoins so explicitly check on isAntiJoins is more clear than check on other join types?

File: velox/exec/HashProbe.cpp (hunk 2 above)
Comment (xiaoxmeng) [reply velox/exec/HashProbe.cpp:288]: Thanks and updated!

File: velox/exec/HashProbe.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Yuhta) [reply velox/exec/HashProbe.cpp]: It's more clear that LEFT and FULL and ANTI without filter in the first early return point, and ANTI with filter at second early return point.

[the remark this answers is not in the document]
Comment (xiaoxmeng) [reply velox/exec/HashProbe.cpp]: Does the following check can cover all the cases?
```
    if (!isAntiJoins(joinType_) || (filter_ == nullptr)) {
      return;
    }
```

[the remark this answers is not in the document]
Comment (Yuhta) [reply velox/exec/HashProbe.cpp]: It's correct either way, just a matter of readability (and not a huge one).  We can do the current way if you want.

**Review by mbasmanova — APPROVED** [review]
@xiaoxmeng Thanks.

**Pull request closed without merging**